File: cpp/34. 在排序数组中查找元素的第一个和最后一个位置.cpp
class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int n = nums.size();
        int left = 0,right = nums.size()-1;
        while(left <= right){
            int mid = (left + right) / 2;
            if (nums[mid] >= target){
                right = mid-1;
            }else {
                left = mid + 1;
            }
        }
        int ind1=left;
        left = 0, right = n - 1;
        while(left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] > target){
                right = mid-1;
            } else {
                left = mid + 1;
            }
        }
        int ind2=left;
        cout<<ind1<<" "<<ind2<<endl;
        if (ind1 <= ind2-1 && ind2-1 < n && nums[ind1] == target && nums[ind2-1] == target)
        return {ind1, ind2-1};
        return {-1,-1};
    }
};

File: 刷题/golang/34. 在排序数组中查找元素的第一个和最后一个位置.go
func searchRange(nums []int, target int) []int {
    n := len(nums)
    ind1 := binarySearch(nums, target, true)
    ind2 := binarySearch(nums, target, false)
    if ind1 <= ind2 - 1 && ind2 - 1 < n && nums[ind1] == target && nums[ind2 - 1] == target {
        return []int{ind1, ind2 - 1}
    }
    return []int{-1, -1}
}
func binarySearch(nums []int, target int, lower bool) int{
    n := len(nums)
    left, right := 0, n-1
    for left <= right {
        mid := (left + right) / 2
        if nums[mid] > target || (lower && nums[mid] >= target) {
            right = mid - 1
        } else {
            left = mid + 1
        }
    }
    return left
}